pegasus/test/test_email_preference_helpers.rb

```diff
@@ -0,0 +1,147 @@
+require_relative '../helper_modules/dashboard'
+require_relative './test_helper'
+require_relative '../helpers/email_preference_helpers'
+require 'timecop'
+
+class EmailPreferenceHelperTest < Minitest::Test
+  def test_upsert_new_email_preference_with_valid_attributes_creates_email_preference
+    row_id = EmailPreferenceHelper.upsert!(
+      email: 'test_valid@example.net',
+      opt_in: true,
+      ip_address: '1.1.1.1',
+      source: EmailPreferenceHelper::ACCOUNT_SIGN_UP,
+      form_kind: nil
+    )
+    assert row_id > 0
+    email_preferences = Dashboard.db[:email_preferences].where(id: row_id)
+    assert_equal 1, email_preferences.count
+
+    email_preference = email_preferences.first
+    assert_equal 'test_valid@example.net', email_preference[:email]
+    assert_equal true, email_preference[:opt_in]
+    assert_equal '1.1.1.1', email_preference[:ip_address]
+    assert_equal EmailPreferenceHelper::ACCOUNT_SIGN_UP, email_preference[:source]
+  end
+
+  def test_upsert_email_preference_with_invalid_email_does_not_create_email_preference
+    begin
+      row_id = EmailPreferenceHelper.upsert!(
+        email: 'amidala@naboo',
+        opt_in: true,
+        ip_address: '1.1.1.1',
+        source: EmailPreferenceHelper::ACCOUNT_SIGN_UP,
+        form_kind: nil
+      )
+    rescue StandardError => error
+      assert_equal 'Email does not appear to be a valid e-mail address', error.message
+      assert_nil row_id
+    else
+      fail "Expected an email validation error."
+    end
```

Comment:
I just realized, this can be rewritten like

```ruby
error = assert_raises 'Expected an email validation error.' do
  # upsert code
end
assert_equal  'Email does not appear to be a valid e-mail address', error.message
#...
```

Your call